src/core/Akka.Streams/Dsl/StreamRefs.cs

```diff
@@ -0,0 +1,729 @@
+            new StreamRefsMaster(system);
+    }
+
+    #endregion
+
+    public sealed class StreamRefSettings
+    {
+        public static StreamRefSettings Create(Config config)
+        {
+            if (config == null) throw new ArgumentNullException(nameof(config), "`akka.stream.materializer.stream-ref` was not present");
+            
+            return new StreamRefSettings(
+                bufferCapacity: config.GetInt("buffer-capacity", 32),
+                demandRedeliveryInterval: config.GetTimeSpan("demand-redelivery-interval", TimeSpan.FromSeconds(1)),
+                subscriptionTimeout: config.GetTimeSpan("subscription-timeout", TimeSpan.FromSeconds(30)));
+        }
+        
+        public int BufferCapacity { get; }
+        public TimeSpan DemandRedeliveryInterval { get; }
+        public TimeSpan SubscriptionTimeout { get; }
+
+        public StreamRefSettings(int bufferCapacity, TimeSpan demandRedeliveryInterval, TimeSpan subscriptionTimeout)
+        {
+            BufferCapacity = bufferCapacity;
+            DemandRedeliveryInterval = demandRedeliveryInterval;
+            SubscriptionTimeout = subscriptionTimeout;
+        }
+
+        public string ProductPrefix => nameof(StreamRefSettings);
+
+        public StreamRefSettings WithBufferCapacity(int value) => Copy(bufferCapacity: value);
+        public StreamRefSettings WithDemandRedeliveryInterval(TimeSpan value) => Copy(demandRedeliveryInterval: value);
+        public StreamRefSettings WithSubscriptionTimeout(TimeSpan value) => Copy(subscriptionTimeout: value);
+        
+        public StreamRefSettings Copy(int? bufferCapacity = null, 
+            TimeSpan? demandRedeliveryInterval = null, 
+            TimeSpan? subscriptionTimeout = null) => new StreamRefSettings(
+            bufferCapacity: bufferCapacity ?? this.BufferCapacity,
+            demandRedeliveryInterval: demandRedeliveryInterval ?? this.DemandRedeliveryInterval,
+            subscriptionTimeout: subscriptionTimeout ?? this.SubscriptionTimeout);
+    }
+
+    /// <summary>
+    /// Abstract class defined serialization purposes of <see cref="SourceRefImpl{T}"/>.
+    /// </summary>
+    internal abstract class SourceRefImpl 
+    {
+        public static SourceRefImpl Create(Type eventType, IActorRef initialPartnerRef)
+        {
+            var destType = typeof(SourceRefImpl<>).MakeGenericType(eventType);
+            return (SourceRefImpl)Activator.CreateInstance(destType, initialPartnerRef);
+        }
+        
+        protected SourceRefImpl(IActorRef initialPartnerRef)
+        {
+            InitialPartnerRef = initialPartnerRef;
+        }
+
+        public IActorRef InitialPartnerRef { get; }
+        public abstract Type EventType { get; }
+    }
+    internal sealed class SourceRefImpl<T> : SourceRefImpl, ISourceRef<T>
+    {
+        public SourceRefImpl(IActorRef initialPartnerRef) : base(initialPartnerRef) { }
+        public override Type EventType => typeof(T);
```

Comment:
Wouldn’t be a single assignment better ?

Reply:
It's used only for deserialization (usually once per instance per node). I don't want to allocate underlying field for it, as Hyperion/Json.NET are able to provide generic type param directly from `SourceRefImpl<T>`'s `T` parameter, so the payload will be smaller.